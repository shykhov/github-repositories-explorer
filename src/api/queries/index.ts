export * from './repository.quiries';
export * from './users.queries';
