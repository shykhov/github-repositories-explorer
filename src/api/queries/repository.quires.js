import gql from 'graphql-tag';

export const FETCH_REPOSTORIES_BY_USER = gql`
  query($login: String!, $repositoryItemsCount: Int!) {
    user(login: $login) {
      id
      login
      repositories(first: $repositoryItemsCount) {
        nodes {
          id
          name
          forkCount
          stargazers {
            totalCount
          }
        }
      }
    }
  }
`;
export const FETCH_REPOSTORIES = gql`
  query($queryString: String!, $repositoryItemsCount: Int!) {
    search(query: $queryString, type: REPOSITORY, first: $repositoryItemsCount) {
      repositoryCount
      edges {
        node {
          ... on Repository {
            name
            descriptionHTML
            stargazers {
              totalCount
            }
            forks {
              totalCount
            }
            updatedAt
          }
        }
      }
    }
  }
`;
