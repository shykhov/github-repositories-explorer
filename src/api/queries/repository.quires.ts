import gql from 'graphql-tag';

export const FETCH_REPOSTORIES = gql`
  query($queryString: String!, $before: String, $after: String, $first: Int, $last: Int) {
    search(query: $queryString, type: REPOSITORY, first: $first, last: $last, before: $before, after: $after) {
      repositoryCount
      pageInfo {
        endCursor
        hasNextPage
        hasPreviousPage
        startCursor
      }
      edges {
        node {
          ... on Repository {
            id
            nameWithOwner
            stargazers {
              totalCount
            }
            forks {
              totalCount
            }
            url
          }
        }
      }
    }
  }
`;
