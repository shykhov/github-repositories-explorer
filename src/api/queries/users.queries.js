import gql from 'graphql-tag';

export const FETCH_USERS = gql`
  query($name: String!, $userItemsCount: Int!) {
    search(query: $name, first: $userItemsCount, type: USER) {
      edges {
        node {
          ... on User {
            id
            login
            name
            avatarUrl
          }
        }
      }
    }
  }
`;
