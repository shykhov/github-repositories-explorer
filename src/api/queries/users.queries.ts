import gql from 'graphql-tag';

export const FETCH_USERS = gql`
  query($name: String!) {
    search(query: $name, first: 100, type: USER) {
      edges {
        node {
          ... on User {
            id
            login
            name
            avatarUrl
          }
        }
      }
    }
  }
`;
