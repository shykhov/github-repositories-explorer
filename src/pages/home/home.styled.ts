import styled from 'styled-components';

export const Root = styled.div`
  margin: 50px;
  display: flex;
`;

export const ControlsWrapper = styled.div`
  width: 400px;
  margin-left: 50px;

  & > div:first-of-type {
    margin-bottom: 25px;
  }
`;
