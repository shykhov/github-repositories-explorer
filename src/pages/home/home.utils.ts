import isPlainObject from 'lodash/isPlainObject';
import isEmpty from 'lodash/isEmpty';
import isNil from 'lodash/isNil';

interface SearchParams {
  [repositorySearchValue: string]: string | null;
  userLogin: string | null;
  page: string | null;
  rowsPerPage: string | null;
}

export const prepareSearchParams = (params: SearchParams): string => {
  if (isPlainObject(params) && !isEmpty(params)) {
    const result = Object.keys(params).reduce((accumulator, currentValue, index) => {
      const newValue = currentValue || '';
      const paramsValue: string | null = params[newValue];
      if (isNil(currentValue) || isNil(paramsValue) || isEmpty(paramsValue)) {
        return accumulator;
      }

      return `${accumulator}${accumulator.length === 1 ? '' : '&'}${currentValue}=${paramsValue || ''}`;
    }, '?');

    return result.length === 1 ? '' : result;
  }

  return '';
};

interface QueryParams {
  [name: string]: string | null;
  owner: string | null;
}

export const prepareQueryParams = (params: QueryParams): string => {
  if (isPlainObject(params) && !isEmpty(params)) {
    const result = Object.keys(params).reduce((accumulator, currentValue) => {
      const newValue = currentValue || '';
      const paramsValue: string | null = params[newValue];
      if (isNil(currentValue) || isNil(paramsValue) || isEmpty(paramsValue)) {
        return accumulator;
      }

      return `${accumulator}${currentValue}:${paramsValue}`;
    }, '');

    return result;
  }

  return '';
};
