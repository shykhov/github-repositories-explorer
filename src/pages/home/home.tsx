import React, { useCallback, useMemo, ChangeEvent, FC } from 'react';
import { useLazyQuery } from '@apollo/react-hooks';
import { ApolloError } from 'apollo-client';
import debounce from 'lodash/debounce';
import { History } from 'history';
import { RouteComponentProps } from 'react-router-dom';

import { Root, ControlsWrapper } from './home.styled';
import { useUrlQuery } from '../../hooks';
import { FETCH_REPOSTORIES, FETCH_USERS } from '../../api/queries';
import {
  REPOSITORIES_PATHNAME,
  REPOSITORIES_PER_PAGE,
  DEFAULT_DEBOUNCE_TIMEOUT,
  USER_LOGIN_PARAMETER,
  REPOSITORY_NAME_QUERY_PARAMETER,
  REPOSITORY_NAME_QUERY_KEY,
  DEFAULT_REPOSITORY_SORT_QUERY,
  PAGINATION_DIRECTION,
} from '../../constants';
import {
  generateCursor,
  formatRepositories,
  formatUsers,
  findCurrentUser,
  prepareQueryString,
  prepareSearchParams,
  RepositoryOptions,
  UserOptions,
  SelectOption,
} from '../../utils';
import { RepositoriesSearchInput } from '../../components/repositories-search-input';
import { RepositoryTable } from '../../components/repository-table';
import { UserSearchSelect } from '../../components/user-search-select';

export type ApolloLazyQueryResult<T> = {
  data: T;
  called: boolean;
  error?: ApolloError | undefined;
  loading: boolean;
};

export interface HomeProps {
  getRepositories(variables: Record<string, unknown>): void;
  repositoriesQuery: ApolloLazyQueryResult<RepositoryOptions>;
  usersQuery: ApolloLazyQueryResult<UserOptions>;
  getUsers(variables: Record<string, unknown>): void;
  history: History;
}

export const Home: FC<HomeProps> = props => {
  const { getRepositories, repositoriesQuery, usersQuery, getUsers, history } = props;
  const query = useUrlQuery();
  const [userLoginSearchValue, setUserLoginSearchValue] = React.useState('');
  const [repositoryNameSearchValue, setRepositoryNameSearchValue] = React.useState('');
  const [page, setPage] = React.useState(0);

  const userLoginParameter = query.get(USER_LOGIN_PARAMETER);
  const repositoryNameSearchParameter = query.get(REPOSITORY_NAME_QUERY_PARAMETER);

  const userOptions = useMemo(() => formatUsers(usersQuery.data), [usersQuery.data]);

  const currentUser = useMemo(() => findCurrentUser(userOptions, userLoginParameter), [
    userOptions,
    userLoginParameter,
  ]);

  const repositoriesData = useMemo(() => formatRepositories(repositoriesQuery.data), [repositoriesQuery.data]);

  const handleChangePage = (event: ChangeEvent, newPage: number) => {
    const {
      repositoryCount,
      pageInfo: { endCursor, startCursor },
    } = repositoriesData;

    const { first, last } = PAGINATION_DIRECTION;

    if (newPage * REPOSITORIES_PER_PAGE + REPOSITORIES_PER_PAGE <= repositoryCount) {
      getRepositories({
        variables: {
          queryString: prepareQueryString({
            [REPOSITORY_NAME_QUERY_KEY]: repositoryNameSearchParameter,
            user: userLoginParameter,
            ...DEFAULT_REPOSITORY_SORT_QUERY,
          }),
          [newPage > page || newPage === 0 ? first : last]: REPOSITORIES_PER_PAGE,
          ...generateCursor({ newPage, oldPage: page, endCursor, startCursor }),
        },
      });
    }

    setPage(newPage);
  };

  const getNewUsers = (userLogin: string) => {
    getUsers({
      variables: {
        name: userLogin,
      },
    });
  };

  const debouncedHandleUserSelectInputChange = useCallback(debounce(getNewUsers, DEFAULT_DEBOUNCE_TIMEOUT), []);

  const handleUserSelectInputChange = (userLogin: string) => {
    if (userLogin) {
      setUserLoginSearchValue(userLogin);
      debouncedHandleUserSelectInputChange(userLogin);
    }
  };

  const handleSubmitInputSearch = (repositoryName: string): void => {
    setPage(0);

    const currentRepositoryName: string = repositoryName || repositoryNameSearchValue;

    history.push({
      pathname: REPOSITORIES_PATHNAME,
      search: prepareSearchParams({
        [REPOSITORY_NAME_QUERY_PARAMETER]: currentRepositoryName,
        [USER_LOGIN_PARAMETER]: userLoginParameter,
      }),
    });
  };

  const debouncedSubmitInputSearch = useCallback(debounce(handleSubmitInputSearch, DEFAULT_DEBOUNCE_TIMEOUT), [
    userLoginParameter,
    repositoryNameSearchParameter,
  ]);

  const handleSearchRepositoryInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    if (value) {
      debouncedSubmitInputSearch(value.trim());
    } else {
      debouncedSubmitInputSearch.cancel();

      history.push({
        pathname: REPOSITORIES_PATHNAME,
        search: prepareSearchParams({
          [USER_LOGIN_PARAMETER]: userLoginParameter,
        }),
      });
    }

    setRepositoryNameSearchValue(value);
  };

  const handleSelectChange = (user: SelectOption) => {
    setPage(0);
    history.push({
      pathname: REPOSITORIES_PATHNAME,
      search: prepareSearchParams({
        [REPOSITORY_NAME_QUERY_PARAMETER]: repositoryNameSearchParameter,
        [USER_LOGIN_PARAMETER]: user ? user.value : '',
      }),
    });
  };

  React.useEffect(() => {
    if (repositoryNameSearchParameter || userLoginParameter) {
      getRepositories({
        variables: {
          first: REPOSITORIES_PER_PAGE,
          queryString: prepareQueryString({
            [REPOSITORY_NAME_QUERY_KEY]: repositoryNameSearchParameter,
            user: userLoginParameter,
            ...DEFAULT_REPOSITORY_SORT_QUERY,
          }),
        },
      });
    }
  }, [getRepositories, userLoginParameter, repositoryNameSearchParameter]);

  return (
    <Root>
      <RepositoryTable
        error={Boolean(repositoriesQuery.error)}
        loading={repositoriesQuery.loading}
        called={repositoriesQuery.called}
        repositoriesData={repositoriesData}
        page={page}
        handleChangePage={handleChangePage}
      />
      <ControlsWrapper>
        <RepositoriesSearchInput
          value={repositoryNameSearchValue || repositoryNameSearchParameter}
          loading={repositoriesQuery.loading}
          handleSearchInputChange={handleSearchRepositoryInputChange}
          debouncedSubmitInputSearch={debouncedSubmitInputSearch}
        />
        <UserSearchSelect
          options={userOptions}
          inputValue={userLoginSearchValue}
          loading={usersQuery.loading}
          onSelectChange={handleSelectChange}
          onInputChange={handleUserSelectInputChange}
          value={currentUser}
        />
      </ControlsWrapper>
    </Root>
  );
};

export const HomeContainer: FC<RouteComponentProps> = ({ history }) => {
  const [getRepositories, repositoriesQuery] = useLazyQuery(FETCH_REPOSTORIES);
  const [getUsers, usersQuery] = useLazyQuery(FETCH_USERS);

  return (
    <Home
      history={history}
      usersQuery={usersQuery}
      getUsers={getUsers}
      repositoriesQuery={repositoriesQuery}
      getRepositories={getRepositories}
    />
  );
};
