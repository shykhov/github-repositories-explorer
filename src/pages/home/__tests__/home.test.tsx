/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { createBrowserHistory, createLocation } from 'history';
import { MockedProvider } from '@apollo/react-testing';
import { match as routerMatch } from 'react-router';

import { FETCH_REPOSTORIES, FETCH_USERS } from '../../../api/queries';
import { Home } from '../home';

jest.mock('react-router-dom', () => ({
  useLocation: jest.fn().mockReturnValue({
    pathname: '/another-route',
    search: '',
    hash: '',
    state: null,
    key: '5nvxpbdafa',
  }),
}));

const mocks = [
  {
    request: {
      query: FETCH_REPOSTORIES,
      variables: {
        queryString: 'in:name react',
        first: 50,
      },
    },
    result: {
      data: {
        search: {
          repositoryCount: 3,
          pageInfo: {
            endCursor: 'Y3Vyc29yOjM=',
            hasNextPage: false,
            hasPreviousPage: false,
            startCursor: 'Y3Vyc29yOjE=',
          },
          edges: [
            {
              node: {
                id: 'MDEwOlJlcG9zaXRvcnkxMTc2ODgyMTY=',
                nameWithOwner: 'ASDQWEA1/ASDQWEA1',
                stargazers: { totalCount: 0 },
                forks: { totalCount: 0 },
                url: 'https://github.com/ASDQWEA1/ASDQWEA1',
              },
            },
            {
              node: {
                id: 'MDEwOlJlcG9zaXRvcnkxNjQ1MTA5MDg=',
                nameWithOwner: 'JFHWYFHEDY9FR7923RYBHFKDSCF/asdqweasdwaeasd',
                stargazers: { totalCount: 0 },
                forks: { totalCount: 0 },
                url: 'https://github.com/JFHWYFHEDY9FR7923RYBHFKDSCF/asdqweasdwaeasd',
              },
            },
            {
              node: {
                id: 'MDEwOlJlcG9zaXRvcnkxNDg0NTcyNzQ=',
                nameWithOwner: 'SLCDevelopment/asdqweasdqweaaa',
                stargazers: { totalCount: 0 },
                forks: { totalCount: 0 },
                url: 'https://github.com/SLCDevelopment/asdqweasdqweaaa',
              },
            },
          ],
        },
      },
    },
  },
  {
    request: {
      skip: false,
      query: FETCH_USERS,
      variables: {
        query: 'gaeron',
      },
    },
    result: {
      data: {
        search: {
          edges: [
            {
              node: {
                id: 'MDQ6VXNlcjUyMzA0NjY=',
                login: 'asdlei99',
                name: 'annlei',
                avatarUrl: 'https://avatars3.githubusercontent.com/u/5230466?v=4',
              },
            },
            {
              node: {
                id: 'MDQ6VXNlcjMzNjU1MTA=',
                login: 'asdlei00',
                name: 'asdlei',
                avatarUrl: 'https://avatars1.githubusercontent.com/u/3365510?v=4',
              },
            },
          ],
        },
      },
    },
  },
];

const history = createBrowserHistory();
const path = `/route/:id`;

const match: routerMatch<{ id: string }> = {
  isExact: false,
  path,
  url: path.replace(':id', '1'),
  params: { id: '1' },
};

const location = createLocation(match.url);

describe('<Home />', () => {
  let renderedComponent: ShallowWrapper;
  const propsMock = { history, match, location };

  beforeEach(() => {
    renderedComponent = shallow(
      <MockedProvider addTypename={false} mocks={mocks}>
        <Home {...propsMock} />
      </MockedProvider>,
    );
  });

  it('should render Home', () => {
    expect(renderedComponent.dive().dive().dive()).toMatchSnapshot();
  });
});
