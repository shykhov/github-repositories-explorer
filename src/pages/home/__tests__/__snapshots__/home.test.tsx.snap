// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render Home 1`] = `
<Home
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/route/1",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": false,
      "params": Object {
        "id": "1",
      },
      "path": "/route/:id",
      "url": "/route/1",
    }
  }
/>
`;
